### DISCUSSION TITLE
Running uPython script from C and calling C functions from the script

### DISCUSSION THREAD

**Discussion by empi91** [micropython/micropython#9854] [Using MicroPython]
Hi, I didn't find anything similar enough in the old topics both here and on forum, so I decided to post a new one, hope that's ok

I have complex C application, where I want to add a functionality of executing external uPython scripts, but at the same time be able to call some defined C functions from this script during the execution. 

So basically something like that:
_C-based script component:_
```C
#include <...>

static  void python_print(char text)
static int python_getParameterValue(char paramID)
static int python_setParameterValue(int paramValue)

int main() {
	Py_Initialize();
	fp = _Py_fopen(file, "r");
	PyRun_SimpleFile(fp, file);
	Py_Finalize();
	return 0;
}
```

_uPython script:_
```py
value1 = 10
text1 = "Hello from Python"
param1_ID = Module1.Name

python_print(text1)
python_getParameterValue(param1_ID)
python_setParameterValue(value1)
```

But while for a standard Python it seems relatively easy:
- add `#include <python.h>`
- link the python library to the C file
- somehow (I'm not there yet) make C functions visible for Python script (create Python modules from those C functions??)

then for uPython it seems more complicated and I got lost. 
Here: [Building MP as static library](https://forum.micropython.org/viewtopic.php?f=3&t=214) and here: [Embedding MP as library in C](https://forum.micropython.org/viewtopic.php?t=705) read I have to make a library on my own

Here: [Writing C modules](https://forum.micropython.org/viewtopic.php?f=2&t=1411) there is something about writing modules in C, but I am unable to put it all together

The relatively easy way to do requires creating a .io library file from C file, which I am not very happy with  (at least that's what I found) so every help, or just a proper direction will be much appreciated!

I did something similar using Lua and it's working, but now just trying to compare it with uPython

**Comment by dlech** [comment]
Did you look at https://github.com/micropython/micropython/tree/master/examples/embedding?

And [this pull request](https://github.com/micropython/micropython/pull/9529) hasn't been merged, but could be of interest.

**Reply by empi91** [reply]
Thank you very much for your answer!

I read through the link you sent, but from what I understand it is more about embedding some uPython commands directly into C code, while what I am looking for is running the external uPython script and pretty much all I need to do in the C code is to start the uPython interpreter and load the script to it (at least that's how it is working in Python. Is uPython different ?)
The code for doing that in C for standard Python would look similar to that:
```py
 	Py_Initialize();

	fp = _Py_fopen(filename, "r");
	PyRun_SimpleFile(fp, filename);

	Py_Finalize();
	return 0;
```

And my question is, is there a similar way of doing it in uPython (would this code work actually?) and if not, is there any workaround which would allow to load and run any uPython script from the C program, without need of changing the C source code?

**Reply by dlech** [reply]
It is just a simple "hello world" style example. You can modify it to suit your needs.

Basically, 

```c
    // Initialized stack limit
    mp_stack_set_limit(40000 * (sizeof(void *) / 4));
    // Initialize heap
    gc_init(heap, heap + sizeof(heap));
    // Initialize interpreter
    mp_init();
```

is the equivalent of `Py_Initialize()`.

and

```c
    // run finalizers
    gc_sweep_all();
    // tear down
    mp_deinit();
```

is the equivalent of `Py_Finalize()`.

In between, you can call just about any MicroPython function. To run a file:

```c
    pyexec_file("filename");
```

might work.

**Reply by empi91** [reply]
It took me a while to went through the code you posted, but another (very) newbie question arise: where in documentation do I find something about all those functions you used (ie. `mp_stack_set_limit`, `pyexec_file`, etc) ? I mean, I did find their definitions in the corresponding files (`runtime.c`, `pyexec.c` etc), but is there anything in the documentation? 

I mean, everywhere I look I was reading uPython is a Python, just a little different, but then even simple interpreter initialization is different, so I'm seeking for some materials to learnt from, to avoid posting newbie questions here :/

**Reply by dlech** [reply]
The C API is not documented, so you just have to look in the source code.

**Reply by empi91** [reply]
Ok, I spent several days digging and trying, I got the DIY custom C module working, but I am unable to get `pyexec_file("filename")` work..
I did `#include "shared/runtime/pyexec.h"` in main C file (working around the `Embedded` example all the time), I added the following lines in `Makefile.upylib`:
```py
SHARED_SRC_C = $(addprefix shared/,\
        libc/printf.c \
        runtime/gchelper_generic.c \
        timeutils/timeutils.c \
        runtime/pyexec.c \
        readline/readline.c \
        $(SHARED_SRC_C_EXTRA) \
        )

(...)

# List of sources for qstr extraction
SRC_QSTR += $(SRC_C) $(SHARED_SRC_C) shared/readline/readline.c shared/runtime/pyexec.c
```
and all I got after trying to `make` is:
```py
In file included from ../../py/lexer.h:31:0,
                 from ../../py/compile.h:29,
                 from ../../shared/runtime/pyexec.c:32:
../../shared/runtime/pyexec.c: In function ‘pyexec_friendly_repl’:
../../py/mpconfig.h:1749:61: error: expected ‘)’ before ‘MICROPY_PLATFORM_COMPILER’
 #define MICROPY_BANNER_MACHINE MICROPY_PY_SYS_PLATFORM " [" MICROPY_PLATFORM_COMPILER "] version"
```

I didn't find anything about such error, code for me looks ok, I kinda give up, so if anyone has an idea (I guess it's some stupid error, but it's been around for 4 days and no idea so far..

**Reply by jimmo** [reply]
> I didn't find anything about such error, code for me looks ok, I kinda give up, so if anyone has an idea (I guess it's some stupid error, but it's been around for 4 days and no idea so far..

Now that you're using pyexec.c it will be trying to use the `MICROPY_BANNER_MACHINE` macro (this comes from mpconfig.h)

```
// String used for the second part of the banner, and sys.implementation._machine
#ifndef MICROPY_BANNER_MACHINE
#ifdef MICROPY_HW_BOARD_NAME
#define MICROPY_BANNER_MACHINE MICROPY_HW_BOARD_NAME " with " MICROPY_HW_MCU_NAME
#else
#define MICROPY_BANNER_MACHINE MICROPY_PY_SYS_PLATFORM " [" MICROPY_PLATFORM_COMPILER "] version"
#endif
#endif
```

I'm guessing you don't have `MICROPY_HW_BOARD_NAME` defined, so it's trying to find definitions for `MICROPY_PY_SYS_PLATFORM` and `MICROPY_PLATFORM_COMPILER` which are likely unset.

**Reply by empi91** [reply]
Oh, thank you, that solved the problem!
However, then there are another 4 errors related to implicit declarations / undeclared functions in `readline.c`, which I am only importing, not changing anything inside. 

And I started to wonder, am I doing something wrong, or basically uPython is just not "working straight out of the box with just slightly modified embedding example from official tree" language?

**Reply by jimmo** [reply]
> And I started to wonder, am I doing something wrong, or basically uPython is just not "working straight out of the box with just slightly modified embedding example from official tree" language?

It does work out of the box if your use case is very simple... but the problem is that when you bring in more functionality, you need to provide the supporting bits. readline means you need to provide stdio. C doesn't really have a better way to do this other than just depending on you to provide the expected methods.

I agree though that the embedding example is pretty useless for four reasons:
 - It's not at all easy to embed in another build system.
 - It's far too minimal and doesn't cover any of the common things that someone is likely to need to do.
 - It assumes an unreasonably high level of knowledge of both C and MicroPython internals.
 - It has a large amount of additional dependency on the unix port, which isn't necessarily useful, but complicates it considerably.

@dlech had already replied by the time I saw your original message, but he linked to exactly what I would have suggested:

> And [this pull request](https://github.com/micropython/micropython/pull/9529) hasn't been merged, but could be of interest.

This is the first step of our plans to improve this situtation. I have been trying to address this for some time (see https://github.com/micropython/micropython/pull/5964 which preceded it). This solves the first point above. The next step is much more detailed examples.

**Reply by empi91** [reply]
Thank you very much for such an extensive answer!
Yes, I went through that example as well, but I think I unfortunately need more classical approach, with being able to execute file with uPython script from C code.

And therefore I think I have to once again ask for your help (hopefully the last time!):
After I fixed the error about lacking `MICROPY_HW_BOARD_NAME ` definition, I'm trying to compile that project, but I got bunch of new errors, related to `readline_hist` this time:
```py
../../shared/readline/readline.c: In function ‘readline_init0’:
../../shared/readline/readline.c:55:12: error: implicit declaration of function ‘MP_STATE_PORT’; did you mean ‘MP_STATE_MEM’? [-Werror=implicit-function-declaration]
     memset(MP_STATE_PORT(readline_hist), 0, MICROPY_READLINE_HISTORY_SIZE * sizeof(const char*));
            ^~~~~~~~~~~~~
            MP_STATE_MEM
../../shared/readline/readline.c:55:26: error: ‘readline_hist’ undeclared (first use in this function); did you mean ‘readline_init’?
     memset(MP_STATE_PORT(readline_hist), 0, MICROPY_READLINE_HISTORY_SIZE * sizeof(const char*));
                          ^~~~~~~~~~~~~
                          readline_init
../../shared/readline/readline.c:340:86: error: ‘readline_hist’ undeclared (first use in this function); did you mean ‘readline_t’?
 f (rl.hist_cur + 1 < MICROPY_READLINE_HISTORY_SIZE && MP_STATE_PORT(readline_hist)[rl.hist_cur + 1] != NULL) {
                                                                     ^~~~~~~~~~~~~
                                                                                      readline_t
../../shared/readline/readline.c:570:27: error: ‘readline_hist’ undeclared (first use in this function); did you mean ‘readline_t’?
         && (MP_STATE_PORT(readline_hist)[0] == NULL
                           ^~~~~~~~~~~~~
                           readline_t
```
And while it clearly stated that `readline_hist`, I have no idea where to declare it..

I found something about declaring it the following way:
```py
#define MICROPY_PORT_ROOT_POINTERS \
    const char *readline_hist[8]; \
```
inside `mpconfigport` file (in case of the `embedding` example it's `mpconfigport_minimal` I guess, but it did not solve the problem..
And idea?

**Reply by jimmo** [reply]
This is a limitation of the embedding example... there's no way to define root pointers without the build process that makes MP_REGISTER_ROOT_POINTER work. I believe this is fixed in the approach in https://github.com/micropython/micropython/pull/9529

You must be looking at an older version of the repo if you're seeing `MICROPY_PORT_ROOT_POINTERS`.

Are you sure you need readline? This should only be necessary if you're enabling the REPL.

**Reply by empi91** [reply]
I don't think I need it actually, but when i removed it from the build, I got another error:
```py
./build/libmicropython.a(pyexec.o): In function `pyexec_friendly_repl':
pyexec.c:(.text+0x47b): undefined reference to `readline'
pyexec.c:(.text+0x591): undefined reference to `readline'
```
So I guess it's required?

I'm working on the repo cloned directly from micropython github last week, so I don't think it is older version

**Reply by jimmo** [reply]
Assuming you're not calling `pyexec_friendly_repl`, then I think this is because you're calling `pyexec_file`, so you're getting everything else from pyexec.c (which includes pyexec_friendly_repl, which calls readline).

The simplest way to solve this is just to provide a dummy implementation of readline (somewhere in your .c code, e.g. the equivalent of hello-embed.c)

```c
int readline(vstr_t *line, const char *prompt) {
    return 0;
}
```

> I'm working on the repo cloned directly from micropython github last week, so I don't think it is older version

There is no occurrence of `MICROPY_PORT_ROOT_POINTERS` in the current revision of the master branch. Did you clone at v1.19.1 ?

**Reply by empi91** [reply]
Yes, 1.19.1
But I guess I was unclear -- I I found somewhere a topic where this info about `MICROPY_PORT_ROOT_POINTERS` was and just copied it together with `const char *readline_hist[8]; \` to check if that would work. So it was not included in the repo after I clone it

Thanks for a tip about adding the dummy implementation, now it at least build up!

**Comment by empi91** [comment]
Hello again, another (newbie) question showed up:
What version (port?) of micropython is the `mp_init()` function using?
Because, as I have to add some custom C modules to my solution, I need to compile them using `make USER_C_MODULES=../../../path/to/user/module`, but I don't know which port to use for it.
I found the 
```py
    // call port specific initialization if any
    #ifdef MICROPY_PORT_INIT_FUNC
    MICROPY_PORT_INIT_FUNC;
    #endif
```

inside `py/runtime.c` but I am not sure how to use it properly

**Reply by dlech** [reply]
> What version (port?) of micropython is the `mp_init()` function using?

It is used by all ports.  Port-specific defines like `MICROPY_PORT_INIT_FUNC` are usually defined in `mpconfigport.h`.

**Reply by empi91** [reply]
Ok, but how do I ensure that the `mp_init()` called in C code in standalone program is using the same port I used to compile with the C module using `make USER_C_MODULES=../../../`?

And in which `mpconfigport.h` do I need to add  `MICROPY_PORT_INIT_FUNC`?
As far as I understand, `mpconfigport.h` is related to some port (say Unix), while `MICROPY_PORT_INIT_FUNC` is related to `mp_init()` function I am calling in some standalone C program. I don't understand where the connection is

**Reply by dlech** [reply]
If you started with the embedding example, then there is a `mpconfigport.h` file there already.
